### PR TITLE
Make a hierarchical menu in navstories

### PR DESCRIPTION
**Pull request by bennyslbs**
The hierachical menu is based on the structure of the stories (permalink).

---
See: https://groups.google.com/forum/#!topic/nikola-discuss/Ry4LpSMY0Bg

### CODE REVIEW

File: v7/navstories/README.md [hunk 1]
```diff
@@ -1,3 +1,34 @@
...
+        #("f", "Foo"),
+    ),
+}
+NAVIGATION_LINKS_POST_NAVSTORIES = {
+    # Format just as NAVIGATION_LINKS, but content included after navstories entries
+}
+```
+
+To exclude a single story from the navstories meny add the following
+metadata `.. hidefromnav: yep` (yep can be anything) to the story.
```
Comment (Kwpolska) [inline v7/navstories/README.md]: `yes` or `True`

File: v7/navstories/README.md [hunk 2]
```diff
@@ -1,3 +1,34 @@
 This very simple plugin throws all the stories to the navigation bar.
 
-Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
+Started as a Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
+
+Changed to map the stories to a hierachical structure in the menu
```
Comment (Kwpolska) [inline v7/navstories/README.md]: Wait, how deep down does this go? Do you realise you can have only one level of submenus?

File: v7/navstories/README.md [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+based on the permalink structure (defaults to the directory structure
+of the posts).
+
+The menu entries inserted by navstories are inserted after entries from `NAVIGATION_LINKS`.
+Entries listed in `NAVIGATION_LINKS_POST_NAVSTORIES` are inserted after navstories entries.
+
+Format of `NAVIGATION_LINKS_POST_NAVSTORIES` is identical to `NAVIGATION_LINKS`.
+
+Sorting and display names in menu can be controlled for top-level entries via `NAVSTORIES_CFG`.
```
Comment (Kwpolska) [inline v7/navstories/README.md]: no example for this

**Comment by bennyslbs** [comment]
Hi,
Thank you for the review.

See my comments inline

2016-10-31 18:17 GMT+01:00 Chris Warrick <notifications@github.com>:

> *@Kwpolska* commented on this pull request.
> ------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > +NAVSTORIES_MAPPING = {
> +    DEFAULT_LANG: (
> +        # Mapping "Toplevel in permalink" to "Visible text"
> +        # The order is as listed here, entries not listed here are included in the end,
> +        # example (remove initial #):
> +        #("b", "Boo"),
> +        #("f", "Foo"),
> +    ),
> +}
> +NAVIGATION_LINKS_POST_NAVSTORIES = {
> +    # Format just as NAVIGATION_LINKS, but content included after navstories entries
> +}
> +```
> +
> +To exclude a single story from the navstories meny add the following
> +metadata `.. hidefromnav: yep` (yep can be anything) to the story.
>
> yes or True
>
I will change to yes, I will also change the code so that no or False gives
the same result as if hidefromnav isn't included, or what is the normal
behavior for metadata with boolean value?

------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > @@ -1,3 +1,34 @@
>  This very simple plugin throws all the stories to the navigation bar.
>
> -Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +Started as a Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +
> +Changed to map the stories to a hierachical structure in the menu
>
> Wait, how deep down does this go? Do you realise you can have only one
> level of submenus?
>
Only one sub-level, i didn't mention that in the README,
But I can include this:
This plugin generates menus and one level of submenus (further sublevels
are
WARNING: Support for submenus is theme-dependent.

------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > @@ -1,3 +1,34 @@
>  This very simple plugin throws all the stories to the navigation bar.
>
> -Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +Started as a Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +
> +Changed to map the stories to a hierachical structure in the menu
> +based on the permalink structure (defaults to the directory structure
> +of the posts).
> +
> +The menu entries inserted by navstories are inserted after entries from `NAVIGATION_LINKS`.
> +Entries listed in `NAVIGATION_LINKS_POST_NAVSTORIES` are inserted after navstories entries.
> +
> +Format of `NAVIGATION_LINKS_POST_NAVSTORIES` is identical to `NAVIGATION_LINKS`.
> +
> +Sorting and display names in menu can be controlled for top-level entries via `NAVSTORIES_CFG`.
>
> no example for this
>
No, it should be NAVSTORIES_MAPPING, and there is an example for that.

**Comment by bennyslbs** [comment]
Hi,
Thank you for the review.

See my comments inline

2016-10-31 18:17 GMT+01:00 Chris Warrick <notifications@github.com>:

> *@Kwpolska* commented on this pull request.
> ------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > +NAVSTORIES_MAPPING = {
> +    DEFAULT_LANG: (
> +        # Mapping "Toplevel in permalink" to "Visible text"
> +        # The order is as listed here, entries not listed here are included in the end,
> +        # example (remove initial #):
> +        #("b", "Boo"),
> +        #("f", "Foo"),
> +    ),
> +}
> +NAVIGATION_LINKS_POST_NAVSTORIES = {
> +    # Format just as NAVIGATION_LINKS, but content included after navstories entries
> +}
> +```
> +
> +To exclude a single story from the navstories meny add the following
> +metadata `.. hidefromnav: yep` (yep can be anything) to the story.
>
> yes or True
>
I will change to yes, I will also change the code so that no or False gives
the same result as if hidefromnav isn't included, or what is the normal
behavior for metadata with boolean value?

------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > @@ -1,3 +1,34 @@
>  This very simple plugin throws all the stories to the navigation bar.
>
> -Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +Started as a Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +
> +Changed to map the stories to a hierachical structure in the menu
>
> Wait, how deep down does this go? Do you realise you can have only one
> level of submenus?
>
Only one sub-level, i didn't mention that in the README,
But I can include this:
This plugin generates menus and one level of submenus (further sub-levels
are mapped to first level submenus).
WARNING: Support for submenus is theme-dependent.

------------------------------
>
> In v7/navstories/README.md
> <https://github.com/getnikola/plugins/pull/183#pullrequestreview-6481593>:
>
> > @@ -1,3 +1,34 @@
>  This very simple plugin throws all the stories to the navigation bar.
>
> -Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +Started as a Proof-of-concept of a ConfigPlugin.  By request on the mailing list.
> +
> +Changed to map the stories to a hierachical structure in the menu
> +based on the permalink structure (defaults to the directory structure
> +of the posts).
> +
> +The menu entries inserted by navstories are inserted after entries from `NAVIGATION_LINKS`.
> +Entries listed in `NAVIGATION_LINKS_POST_NAVSTORIES` are inserted after navstories entries.
> +
> +Format of `NAVIGATION_LINKS_POST_NAVSTORIES` is identical to `NAVIGATION_LINKS`.
> +
> +Sorting and display names in menu can be controlled for top-level entries via `NAVSTORIES_CFG`.
>
> no example for this
>
No, it should be NAVSTORIES_MAPPING, and there is an example for that.

**Comment by bennyslbs** [comment]
Previous mail was sent twice by an failure, except a small fix, so the last
one had this fix:
This plugin generates menus and one level of submenus (further sub-levels
are mapped to first level submenus).

Br,
Benny
​

**Comment by Kwpolska** [comment]
Please respond on GitHub, not via e-mail. As for the `hidefromnav` variable, no need for `no` or `False` support (check presence, not value)

**Comment by bennyslbs** [comment]
I have updated README.md
Code checking hidefromdev are checking for presence, so nothing changed
(I have also removed my first comment of the dublets).

**Comment by bennyslbs** [comment]
I have installed v. 7.8.1 and found that default location for stories have been changed to pages, commit df8d5d2 made navstories work well for both pages and stories.

Commit 5667017 lets navstories only add menu entries if initial path of permalink match an entry in NAVSTORIES_PATH.

I think that was all for now.

File: v7/navstories/navstories.py [hunk 4]
```diff
@@ -38,12 +38,50 @@ def set_site(self, site):
         site.scan_posts()
         # NAVIGATION_LINKS is a TranslatableSetting, values is an actual dict
         for lang in site.config['NAVIGATION_LINKS'].values:
+            # Which paths are navstories active for for lang?
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: one “for”

File: v7/navstories/navstories.py [hunk 5]
```diff
@@ -38,12 +38,50 @@ def set_site(self, site):
...
             new = []
+            newsub = {}
             for p in site.pages:
-                if lang in p.translated_to and not p.meta('hidefromnav'):
-                    new.append(p)
-            new_entries = tuple(sorted([(p.permalink(lang), p.title(lang)) for p in new]))
-
+                permalink = p.permalink(lang)
+                for s in navstories_paths:
+                    if permalink.startswith('/' + s + '/'):
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: `startswith` can take a tuple.

```python
navstories_paths = tuple('/' + s + '/' for s in navstories_paths)  # above
...
if permalink.startswith(navstories_paths):
```

File: v7/navstories/navstories.py [hunk 6]
```diff
@@ -38,12 +38,50 @@ def set_site(self, site):
...
+                        navpath = permalink[2 + len(s):].split('/') # Permalink format '/A/B/' for a story in s/A/B.rst
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: What do you mean here with the comment?

File: v7/navstories/navstories.py [hunk 7]
```diff
@@ -38,12 +38,50 @@ def set_site(self, site):
...
+                                newsub[navpath[0]].append((p.permalink(lang), p.title(lang)))
+            new_all = [(p.permalink(lang), p.title(lang)) for p in new]
+            for k in sorted(newsub.keys()):
+                new_all.append(tuple((tuple(newsub[k]), k)))
+            new_entries = []
+            navstories_mapping = ()
+            if 'NAVSTORIES_MAPPING' in site.config and lang in site.config['NAVSTORIES_MAPPING']:
+                navstories_mapping = site.config['NAVSTORIES_MAPPING'][lang]
+            for sk, sv in navstories_mapping:
+                for i in range(len(new_all)):
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: This code looks like a mess. Could you make it more readable/understandable?

(Also, I’d drop that setting, it makes people too lazy)

File: v7/navstories/navstories.py (hunk 4 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: I'l fix it

File: v7/navstories/navstories.py (hunk 5 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: Yes, but next I have to remove `'/' + s + '/'`, I don't know how to do it without regular expressions, and navstories_paths are not regular expressions.

File: v7/navstories/navstories.py (hunk 6 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: It was just a note when I started chaing the plugin, I will remove it.

**Comment by bennyslbs** [comment]
I have not a public demo site at the moment, I will create one.

File: v7/navstories/navstories.py (hunk 7 above)
Comment (Kwpolska) [inline v7/navstories/navstories.py]: 1. You could use TranslatableSetting on all your setting to make things less of a mess.
2. Do the rename. And add some comments to explain your `range(len())` thing.
3. Aren’t you using `post.title()`? My point is, if you need that much customization for things (so page titles need to be different in nav and page), perhaps you’re better off making menus by hand, will be much simpler than messing with this plugin.

File: v7/navstories/navstories.py (hunk 5 above)
Comment (Kwpolska) [inline v7/navstories/navstories.py]: Make the check using a tuple, then use a for loop to figure out the **longest** string that the permalink starts with.

Or perhaps:

```py
navstories_paths = tuple(('/' + s.strip('/') + '/') for s in navstories_paths)

for p in site.pages:
    permalink = p.permalink()
    s_candidates = [s for s in navstories_paths if permalink.startswith(s)]
    if not s_candidates:
        continue
    # get longest path
    s = max(s_candidates, key=len)

    navpath = permalink[len(s):].split('/')
```

Use case:

```py
NAVSTORIES_PATHS = {DEFAULT_LANG: ('pages', 'pages/foo', 'pages/foo/bar')}
```

File: v7/navstories/navstories.py (hunk 7 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: 1: Is there a example on using TranslatableSetting, I have searched the plugins repository for it, but didn't find much help.
2: I will do that
3: I am using post.title(), but would like to get a deeper menu
I have created a demo site. Nav and A menu entries are autogenerated. A is not listed in NAVSTORIES_MAPPING, so it comes after Nav.
https://slbs.dk/n/
The conf.py and directory struture under src/  is here: https://slbs.dk/n/listings/

File: v7/navstories/navstories.py (hunk 5 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: I will change to your code.
Should p.permalink() not be p.permalink(lang)?

Comment (Kwpolska) [inline v7/navstories/navstories.py]: Sure.

File: v7/navstories/navstories.py (hunk 7 above)
Comment (Kwpolska) [inline v7/navstories/navstories.py]: Look at nikola.py in core for some samples, and at the utils.TranslatableSetting class (which is well-documented).

**Comment by bennyslbs** [comment]
**p.permalink(lang)**
When p.permalink(lang) shall be changed to p.permalink(), what then with p.title(lang)?

**Usage of TranslatableSetting**
I have looked into nikola.py and utils.py, but I am not sure I understand how you would like it implemented. I will soon commit a proposal.

**Comment by Kwpolska** [comment]
> p.permalink(lang)

Please use the `lang` argument.

> TranslatableSetting

You can use those to avoid some checks when it comes to languages in settings.

```py
navstories_paths = utils.TranslatableSetting('navstories_paths', site.config['NAVSTORIES_PATHS'], site.config['TRANSLATIONS'])
```

You can use `navstories_paths(lang)` to get a translation for `lang`, or default to the default language. It follow the behavior of NAVIGATION_LINKS and other built-in translatable settings.

**Comment by bennyslbs** [comment]
I have updated https://slbs.dk/n/ with a bit in english, and I found out how p.permalink() and p.title(lang) worked together with a multi-language site - Nice :-)

You wrote: "undo this!", I assume that it is not commit bd62af2, since this change is required.
What would you like to get reverted?
I think the first loop over the configuration variables is the smartest way to accept:
- Missing configuration
- Variable created without translation support
- Variable created with translation support
It is similar to what was done in nikola.py

The loop in the top op the lang loop could be skipped, but I think it is easier to read.

**Comment by bennyslbs** [comment]
I had not seen your previous comment - I have just incorporated it, but reading config variables are in a try-excpet to prevent errors if a variable is missed in conf.py.

**Comment by Kwpolska** [comment]
How does your plugin behave when `/nav/index.html` and `/nav/last/index.html` are in output?

**Comment by bennyslbs** [comment]
I am not sure if I understand your question about  `/nav/index.html` and `/nav/last/index.html` are in output?
It ignores all files in the output folder.

I will add text similar to this to the README.md:
Navstories seaches only for pages with permalinks listed in `NAVSTORIES_PATHS`.
It skips the path that matches `NAVSTORIES_PATHS`. Next level of the permalink is used in the top level menu. If there are further levels it maps to sub-menus.
----

I don't know if that answers your question.

**Comment by Kwpolska** [comment]
But how does it react to `/stories/nav.rst` and `/stories/nav/last.rst`? Where is `nav.rst` displayed and how?

**Comment by bennyslbs** [comment]
That will not be handled good, it will give two entries in the main menu, one with submenu for all entries in `nav/*.rst` and one with `nav.rst`.
As I see it it is not possible to make the mapping well, but one way could be moving the top level `nav.rst` to first element in submenu, e.g.:
A menu like tihis
```
  nav
   |- * (`nav.rst`)
   |- Page 1 (`nav/1.rst`)
   |- Page 2 (`nav/2.rst`)
```

**Comment by bennyslbs** [comment]
Last commit 88ac1a8 changes handling of entries as e.g. `nav/*.rst` and one with `nav.rst`.
I have changed it a bit compared to the idea I have listed above:
```
  nav
   |- Top (nav.rst)
   |- * Page 1 (nav/1.rst)
   |- * Page 2 (nav/2.rst)
   |- * * Page 2a (nav/2/a.rst)
```
The indention '* ' can be controlled via conf.py.

**Comment by bennyslbs** [comment]
I don't think I have more changes to the plugin.

File: v7/navstories/navstories.py [hunk 8]
```diff
@@ -26,24 +26,146 @@
...
+            for map_key, map_txt in nav_conf_lang['NAVSTORIES_MAPPING']:
+                # Loop over all new_raw entries, checking if it matches map_key; if match: add it and delete from new_raw
+                if map_key in new_raw:
+                    new.append([map_txt, new_raw[map_key]])
+                    del(new_raw[map_key])
+            # Add remaing new_raw entries which didn't match any map_key
+            new.extend([[None, new_raw[_]] for _ in sorted(new_raw)])
 
+            # Map to tuble
+            new_entries = self.map_to_menu(new)
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: Wouldn’t an OrderedDict work better for `new`?

File: v7/navstories/navstories.py [hunk 9]
```diff
@@ -26,24 +26,146 @@
...
+        """
+        ret = []
+        for topent in entries:
+            # Determine toplevel menu name
+            # - If not None, use the name
+            # - If None, use title of page if page with navpath length=1 exist, else navpath[0]
+            title = topent[0]
+            if not title:
+                # Default is 1th level of navpath
+                title = topent[1][0][0][0]
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: Why one, zero, zero, zero? Those indexes are really hard to follow.

That said, you could get rid of at least one by doing

```py
for title, topent_one in entries:
```

(come up with better name for `topent_one`), or perhaps OrderedDict (as noted below).

File: v7/navstories/navstories.py [hunk 10]
```diff
@@ -26,24 +26,146 @@
...
+          - List
+            - Top level Menu text, or None if auto-mapped, i.e. not in NAVSTORIES_MAPPING
+            - List of pages in the top menu entry
+              - List of:
+                - navpath: List containing navigation hieracy (permalink without langinfo (initial /en/) and without the NAVSTORIES_PATHS (e.g. /pages/))
+                - Permalink
+                - Page title
+        Example:
+        [   [   'Menu text for nav',
+                [   [['nav'], '/pages/nav/', 'nav/'],
```
Comment (bennyslbs) [inline v7/navstories/navstories.py]: NavNode is a class that you suggests that I create?
I think NavNode should contain:
Member variables:
- hierachy
- permalink
- title
And maybe a constructor method for creating the content based on the page variable (p)

File: v7/navstories/navstories.py (hunk 9 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: I will take a look on that.

File: v7/navstories/navstories.py [hunk 11]
```diff
@@ -26,24 +26,146 @@
...
+                for i in range(len(topent[1])):
+                    if len(topent[1][i][0]) == 1:
+                        title = topent[1][i][2] # Page Title
+            if len(topent[1]) == 1 and len(topent[1][0][0]) == 1:
+                # Only one menu item and it is not a subpage, let the item go direct to top level menu
+                #ret += tuble(topent[1][0][1], title)
+                pass
+            else:
+                sub = []
+                # Find min/max depth in actual submenu
```
Comment (bennyslbs) [inline v7/navstories/navstories.py]: Done :-)

File: v7/navstories/navstories.py [hunk 12]
```diff
@@ -26,24 +26,146 @@
...
+                max_depth = 0 # Low initial value, set to max(len(navpath)) in this submen
+                for p in topent[1]:
+                    min_depth = min(min_depth, len(p[0]))
+                    max_depth = max(max_depth, len(p[0]))
+                # Map pages to submenu
+                for p in sorted(topent[1], key=lambda x: x[1]): # Sort by second element (permalink) in page list
+                    prefix = self.navstories_submenu_indention * (len(p[0]) - min_depth)
+                    sub.append(tuple([p[1], prefix + p[2]]))
+                ret.append(tuple([tuple(sub), title]))
+        return tuple(ret)
```
Comment (bennyslbs) [inline v7/navstories/navstories.py]: No, it returns a tuple of the same format as  NAVIGATION_LINKS (appended to NAVIGATION_LINKS as the last thing that navstories do)

File: v7/navstories/navstories.py [hunk 13]
```diff
@@ -26,24 +26,146 @@
...
+                nav_config[i] = utils.TranslatableSetting(i, {}, site.config['TRANSLATIONS'])
+
         site.scan_posts()
         # NAVIGATION_LINKS is a TranslatableSetting, values is an actual dict
         for lang in site.config['NAVIGATION_LINKS'].values:
-            new = []
+            # navstories config for lang
+            nav_conf_lang = {}
+            for i in self.conf_vars:
+                 nav_conf_lang[i] = nav_config[i].values[lang]
```
Comment (bennyslbs) [inline v7/navstories/navstories.py]: I have seen ```.values[lang]``` have been used elsewhere, is the result the same of ```nav_config[i].values[lang]``` and ```nav_config[i](lang)```?

Comment (Kwpolska) [inline v7/navstories/navstories.py]: `.values` is internal use and will break if someone assumes the usual behaviour of using default value if language is untranslated to.

File: v7/navstories/navstories.py (hunk 12 above)
Comment (Kwpolska) [inline v7/navstories/navstories.py]: PS. you misspell `tuple` a lot in your code, so `s/tuble/tuple/` please.

File: v7/navstories/navstories.py (hunk 13 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: Should ```site.config['NAVIGATION_LINKS'].values[lang]``` also be changed?
And what about ```for lang in site.config['NAVIGATION_LINKS'].values:```?

Comment (Kwpolska) [inline v7/navstories/navstories.py]: `for` → can’t be changed, not iterable
other instances where there is no iteration: go ahead

File: v7/navstories/navstories.py (hunk 8 above)
Comment (bennyslbs) [inline v7/navstories/navstories.py]: I will change new to a OrderedDict where the key is the title and the value is an array of NavNode's.

**Comment by bennyslbs** [comment]
I have implemented the changes, should I increment the version number again, or is it only when on master it should be incremented?

**Comment by Kwpolska** [comment]
Those version numbers don’t matter.

File: v7/navstories/navstories.py [hunk 14]
```diff
@@ -26,24 +26,160 @@
...
+                s_candidates = [s for s in paths if permalink_nolang.startswith(s)]
+                if not s_candidates:
+                    continue
+                # get longest path
+                s = max(s_candidates, key=len)
+                # Strip off the longest path in paths
+                navpath = permalink_nolang[len(s):].strip('/').split('/')
+                if len(navpath) == 0:
+                    # Should not happen that navpath is empty, but to prevent errors, and inform via a warning
+                    LOGGER.warn("Page with permalink: '%s', title: '%s', not added to menu by navstories.\033[0m" % (p.permalink(lang), p.title(lang)))
```
Comment (Kwpolska) [inline v7/navstories/navstories.py]: Nikola will take care of ending formats itself. (I’ll fix this one.)

**Comment by Kwpolska** [comment]
Thank you for contributing to Nikola! :tada:
